charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,302 @@
+
+    def _check_and_report_k8s_conflict(self, error) -> bool:
+        """Return True if error status code is 409 (conflict), False otherwise."""
+        if error.status.code == 409:
+            self.logger.warning(f"Encountered a conflict: {error}")
+            return True
+        return False
+
+    def _apply_k8s_resources(self, force_conflicts: bool = False) -> None:
+        """Apply K8s resources.
+
+        Args:
+            force_conflicts (bool): *(optional)* Will "force" apply requests causing conflicting
+                                    fields to change ownership to the field manager used in this
+                                    charm.
+                                    NOTE: This will only be used if initial regular apply() fails.
+        """
+        self.unit.status = MaintenanceStatus("Creating K8s resources")
+        try:
+            self.k8s_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying K8s resources
+                # re-apply K8s resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying K8s resources")
+                self.logger.warning("Apply K8s resources with forced changes against conflicts")
+                self.k8s_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("K8s resources creation failed") from error
+        try:
+            self.crd_resource_handler.apply()
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying CRD resources
+                # re-apply CRD resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying CRD resources")
+                self.logger.warning("Apply CRD resources with forced changes against conflicts")
+                self.crd_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("CRD resources creation failed") from error
+        self.model.unit.status = MaintenanceStatus("K8s resources created")
 
-    def _check_leader(self):
+    def _check_leader(self) -> None:
+        """Check whether a unit is the leader."""
         if not self.unit.is_leader():
             # We can't do anything useful when not the leader, so do nothing.
-            raise CheckFailed("Waiting for leadership", WaitingStatus)
+            self.logger.warning("Not a leader, skipping setup")
+            raise ErrorWithStatus("Waiting for leadership", WaitingStatus)
+
+    def _check_container_connection(self) -> None:
+        """Check if connection can be made with container."""
+        if not self.container.can_connect():
+            raise ErrorWithStatus("Pod startup is not complete", MaintenanceStatus)
```

Comment:
Do we use `MaintenanceStatus` for something like this in other charms?  I wouldn't have thought `Maintenance` here, but it could work.  I think `Maintenance` is typically when we know we're doing something - I'd have used `WaitingStatus`

Reply:
Although `Waiting` does feel a bit more like "waiting on something outside my control", whereas `Maintenance` could be waiting on something I'm doing

Reply:
I believe [`Maintenance`](https://ops.readthedocs.io/en/latest/index.html?highlight=maintenancestatus#ops.model.MaintenanceStatus) is the right one here as the unit is doing some preparation work (in this case, the container is spinning up), while [`Waiting`](https://ops.readthedocs.io/en/latest/index.html?highlight=maintenancestatus#ops.model.WaitingStatus) is more of an external process that should happen before the unit can progress to an active state.

Reply:
> Do we use MaintenanceStatus for something like this in other charms? 

I'd say we do more often than not, e.g. in the KFP API Server, Jupyter, or Seldon charms.

Reply:
:+1: